### PR TITLE
feat(build): add option to build specific dir

### LINKED ISSUE #8071: Compile a subdirectory
### Component

Forge

### Describe the feature you would like

While working invariant testing projects with Echidna/Medusa, I often encounter a compilation issue related to Foundry's default project structure.

By default, Echidna/Medusa uses [crytic-compile](https://github.com/crytic/crytic-compile) to compile a project using Foundry if `.` is passed as the target. This results in a [`forge build --build-info --skip */test/** */script/** --force`](https://github.com/crytic/crytic-compile/blob/master/crytic_compile/platform/foundry.py#L57-L73) command.

The problem is that this makes it impossible to host invariant tests under `test/invariants` and compile them using Foundry. The only alternative is to pass `--foundry-compile-all` to crytic-compile, which will then fall back to a `forge build --build-info` command. However, this will compile ALL Solidity files, which is also undesired, since in most cases the majority of files are unit and fuzz test files.

This results in a significant slowdown of the invariant testing process, and forces security researchers to either move invariant tests to `src/`, which is semantically incorrect, or to compile everything and waste time.

This feature request suggests introducing a configuration parameter in Foundry that allows specifying a subdirectory for compilation. This can be achieved by extending the forge build command with an option to compile only the specified subdirectory, such as `forge build --dir test/invariants`. This way, we could either update crytic-compile or simply pass `--ignore-compile` to these tools, relying on Foundry's subdirectory compilation.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
Closes #8071 

Not too familiar with build but at a quick look seemed to be a low hanging fruit with big impact for other tools integration, so here it is, hope it makes sense :smile: 

- Adds capabilities to build specific directories and file (no toml config) by specifying `--paths` switch to `forge build` e.g.
```
forge build --paths test/regressions/ test/invariant/Invariant.sol
```
- builds child dirs within specified dirs
- works in conjunction with `--skip` as well

### CODE REVIEW

File: crates/forge/bin/cmd/build.rs [hunk 1]
```diff
@@ -80,7 +81,22 @@ impl BuildArgs {
...
+            for dir in dirs {
+                for entry in fs::read_dir(dir)? {
+                    let entry = entry?;
+                    let path = entry.path();
+                    if path.is_file() {
+                        files.push(entry.path());
+                    }
+                }
+            }
+        }
```
Comment (mattsse) [inline crates/forge/bin/cmd/build.rs:96]: I believe there should be utils to find the sol/vy files in a dir @klkvr ?

I believe
https://github.com/foundry-rs/compilers/blob/43da903df7231bf60377c85a6d76ae6b088c723d/src/utils.rs#L113-L113

File: crates/cli/src/opts/build/core.rs [hunk 2]
```diff
@@ -126,6 +126,13 @@ pub struct CoreBuildArgs {
     #[serde(skip)]
     pub skip: Option<Vec<SkipBuildFilter>>,
 
+    /// Build files within specific directories.
+    ///
+    /// Child directories are not accounted and should be explicitly added.
+    #[arg(long, num_args(1..))]
+    #[serde(skip)]
+    pub dirs: Option<Vec<PathBuf>>,
```
Comment (mattsse) [inline crates/cli/src/opts/build/core.rs:134]: should we also support individual files?
maybe we do `files:`?

**Review by mattsse — CHANGES_REQUESTED** [review]
iirc a lot of folks are also requesting this, and this actually looks simple to add. because this is opt-in I belive this is fine

ptal @klkvr

**Comment by DaniPopes** [comment]
Can't we just pass these as positional arguments?

File: crates/forge/bin/cmd/build.rs (hunk 1 above)
Comment (klkvr) [reply crates/forge/bin/cmd/build.rs:96]: Right, similarly to this https://github.com/foundry-rs/foundry/blob/b7dcf4661ddb4fa0cf3519c800df2d5c96500a7e/crates/forge/bin/cmd/test/mod.rs#L219

**Comment by klkvr** [comment]
perhaps we also want a config key for this? similar to `skip` from https://github.com/foundry-rs/foundry/pull/8061

File: crates/forge/bin/cmd/build.rs (hunk 1 above)
Comment (grandizzy) [reply crates/forge/bin/cmd/build.rs]: nice, changed it in https://github.com/foundry-rs/foundry/pull/8149/commits/6da9a55dfb23d7202ff0018d8441f19fe1c37efc

File: crates/cli/src/opts/build/core.rs (hunk 2 above)
Comment (grandizzy) [reply crates/cli/src/opts/build/core.rs]: yeah, since now it supports single files I renamed it to `paths`

**Comment by grandizzy** [comment]
> perhaps we also want a config key for this? similar to `skip` from #8061

I think that would be cool, a difference from skip is that the arg is named `paths` in BuildArgs but probably the config key should be meaningful named `build-paths`? Also not sure but would this imply a change in `MultiCompiler`, I could be wrong but does `forge build` command uses only solc compiler?

**Review by mattsse — APPROVED** [review]

**Comment by klkvr** [comment]
> I think that would be cool, a difference from skip is that the arg is named paths in BuildArgs but probably the config key should be meaningful named build-paths? Also not sure but would this imply a change in MultiCompiler, I could be wrong but does forge build command uses only solc compiler?

This wouldn't require a change in compilers, should be doable by only using sparse output filters

> LE: on a 2nd thought, since this cannot be shared with other commands and kind of valid only in build context, do we want to move it as a config key?

yeah I meant making it valid in all context if specified in config. Since https://github.com/foundry-rs/foundry/pull/8061 we respect `skip` for all context by converting it into sparse output filter directly in `Config::create_project`: https://github.com/foundry-rs/foundry/blob/b7dcf4661ddb4fa0cf3519c800df2d5c96500a7e/crates/config/src/lib.rs#L811-L813

**Comment by grandizzy** [comment]
> 
> yeah I meant making it valid in all context if specified in config. Since #8061 we respect `skip` for all context by converting it into sparse output filter directly in `Config::create_project`:
> 
> https://github.com/foundry-rs/foundry/blob/b7dcf4661ddb4fa0cf3519c800df2d5c96500a7e/crates/config/src/lib.rs#L811-L813

I think I got it now, but wouldn't this be an issue with the sparse output filter when we want to include files (instead excluding them?
Let's say I specify file `test/Bar.sol` is `Test` as path to build, then I create a `FileFilter` and implement `is_match` function that would return true if `test/Bar.sol` is a match (but false for `Test`), is there a way to figure out all other deps to build (as rn I don't see the `test.sol` generated in `out` dir, while for `forge build --paths test/Bar.sol` I get them both)

**Comment by klkvr** [comment]
> I think I got it now, but wouldn't this be an issue with the sparse output filter when we want to include files (instead excluding them?
Let's say I specify file test/Bar.sol is Test as path to build, then I create a FileFilter and implement is_match function that would return true if test/Bar.sol is a match (but false for Test), is there a way to figure out all other deps to build (as rn I don't see the test.sol generated in out dir, while for forge build --paths test/Bar.sol I get them both)

Right, the way sparse output works is that it only requests artifacts for matched sources, excluding everything else. Such approach reduces compile time but might be a bit unexpected if people are expecting artifacts for imported sources to also be produced.

However another issue is if we approach this by just using sparse output filters then this wouldn't give users a way to compile files which are not under `src/` as they won't be included into input sources set.

Let's keep the config option for a future follow-up

File: crates/forge/bin/cmd/build.rs [hunk 3]
```diff
@@ -18,6 +18,9 @@ use watchexec::config::{InitConfig, RuntimeConfig};
 
 foundry_config::merge_impl_figment_convert!(BuildArgs, args);
 
+// Extensions accepted by `forge build`
+const BUILD_EXTENSIONS: &[&str] = &["sol", "yul", "vy", "vyi"];
```
Comment (klkvr) [inline crates/forge/bin/cmd/build.rs:22]: can we use `MultiCompilerLanguage::EXTENSIONS` here? It would require including `Language` trait from compilers

Comment (grandizzy) [reply crates/forge/bin/cmd/build.rs]: yep, changed in https://github.com/foundry-rs/foundry/pull/8149/commits/0e71ca5e9c2e12b37696a9261cc7bda999076918

**Review by klkvr — APPROVED** [review]

**Comment by klkvr** [comment]
>> Can't we just pass these as a list of 0.. positional arguments?

> hope I get this comment right, pls check https://github.com/foundry-rs/foundry/commit/6da9a55dfb23d7202ff0018d8441f19fe1c37efc#diff-f3c6bdb4ebdb62f63412254e4c9154aff947fc45daf6aaff880aaf6470182d8aR129-R132
https://github.com/foundry-rs/foundry/pull/8158

@grandizzy sorry missed this, I think Dani meant changing it to positional args to allow `forge build file.sol` commands. Created small follow-up https://github.com/foundry-rs/foundry/pull/8158, just moves `paths` to `BuildArgs` basically

**Pull request merged**